### ISSUE TITLE
Help?

### ISSUE THREAD

**Issue by SpencerCDixon** [gschier/hemlock#1]
This project looks really cool!  Any PR's you'd be open to?  Version command? Docs? etc.

**Comment by gschier** [comment]
Hey @SpencerCDixon,

I appreciate the offer. At this point, the codebase is still changing too much to spend a _lot_ of time on docs. I've mostly been using them to organize my thoughts so far. There may be a few things you can help with, though.

Feedback is probably the biggest thing. I'd love to hear your thoughts after having tried it out.

By "Version" command, do you mean listing the version number via the CLI? If so, that'd be useful. The project doesn't currently maintain a version number, and I'm not sure what the best-practice is for Go libraries, but if you wanted to figure that out that would be awesome.

I'll try to post some issues as things start to stabilize a bit more over the next week.

Thanks!

**Comment by SpencerCDixon** [comment]
Sure I can take a stab at putting together a `version` command.  I generally see folks injecting it at build time using `-ldflags`

For example if you wanted to use the date:
```sh
go build -ldflags "-X main.version=`date -u +.%Y%m%d.%H%M%S`" main.go
```

Probably less important while it's in such active development but I can get it set up anyways so it's there for you later if/when you need it.

I was hoping to play around with the framework this evening when I get home from a trip.. I'll jot some notes down when I do to get you some feedback!

**Comment by gschier** [comment]
Hmm, the `ldflags` thing is interesting but I think it need to be a Git tag in order to be picked up by package managers like [dep](https://golang.github.io/dep/) doesn't it?

Let me know what you think 😄

**Comment by SpencerCDixon** [comment]
Yeah we can totally use a git tag!  In the past I've used git tags, git sha's, and even just hard coded version strings (like you'd do in a package.json) before publishing.

For releasing I'd recommend checking out [goreleaser](https://goreleaser.com/).  It makes generating the binaries really easy and deploying them to github's release page + homebrew + other linux package managers.  [Here is an example of releasing with goreleaser if it helps](https://github.com/SpencerCDixon/exocortex/blob/master/.goreleaser.yml)

One thing I noticed yesterday when I was getting things set up is in the readme you have people `go get` the code but that doesn't actually install the binary, it just fetches the go code.  For someone to be able to run `hemlock new` they first need to do: `go install github.com/gschier/hemlock/cmd/hemlock`.  That command runs into issues if users don't have the required deps installed though.  So using `dep` maybe the move is to `go get` it, then cd into the dir and `dep ensure` it and then finally `go install` it.   Not a big deal atm since project is in active dev but thought I'd let you know!

The starter project that gets pulled down is on an older version that doesn't support struct JSON rendering (or maybe I was doing something wrong ha).  Curious on your thoughts of the pull a project down vs. keeping the starter project with source as a template.  Personally, I think it will be easier to manage the starter template if it's bundled in with the binary and the repo of the core framework.  You can actually bundle in the template files with the binary pretty easily using a tool called [packr](https://github.com/gobuffalo/packr).

**Comment by gschier** [comment]
That's great info! 

I was looking at goreleaser a while back. Looks like there's also a [godownloader](https://github.com/goreleaser/godownloader) to make installation easier as well. I was actually thinking of splitting the CLI into it's own repo since it's not really related to the framework.

It's strange that pulling down the starter project didn't work for you. The framework is checked into the repo inside the `vendor` directory so it should always be in sync. It's possible I checked in a bad commit by accident though since things are moving so quickly.

**Comment by SpencerCDixon** [comment]
If you do a `new` and go into the vendor and look at the code this is what the Data Result looks like:

```go
func (r *Result) Data(data interface{}) interfaces.Result {
	r.flushHeaders()

	r.hasSentData = true

	if v, ok := data.(string); ok {
		r.w.Write([]byte(v))
	} else if v, ok := data.([]byte); ok {
		r.w.Write(v)
	} else if v, ok := data.(io.Reader); ok {
		io.Copy(r.w, v)
	} else {
		v := []byte(fmt.Sprintf("%v", data))
		r.w.Write(v)
	}

	return r
}
```

On your master branch it looks like:
```go
func (r *Result) Data(data interface{}) interfaces.Result {
	// If it's an error, call Result.Error instead
	if err, ok := data.(error); ok {
		return r.Error(err)
	}

	r.flushHeaders()

	r.hasSentData = true

	if v, ok := data.([]byte); ok {
		r.w.Write(v)
	} else if v, ok := data.(io.Reader); ok {
		io.Copy(r.w, v)
	} else {
		dataType := reflect.TypeOf(data)
		if dataType.Kind() == reflect.Ptr {
			dataType = dataType.Elem()
		}

		if dataType.Kind() == reflect.Struct {
			e := json.NewEncoder(r.w)
			e.Encode(data)
		} else {
			v := []byte(fmt.Sprintf("%v", data))
			r.w.Write(v)
		}
	}

	return r
}
``` 

Looks like master does the proper reflection on the struct but the checked in vendor version of the framework does not.

Re: godownloader

> It's a go program that reads a YAML file that uses a template to make a posix shell script.

Haha love it 😄 .  Def worth looking into!  `goreleaser` has worked great for me, think TJ uses it in `up` too.

**Comment by SpencerCDixon** [comment]
Re: CLI being in a separate repo... I actually think that would be somewhat of an anti pattern in the Go world/end up being more of a headache.  It's totally normal/encouraged to build separate binaries for diferent purposes.  [See item number two in this blog post](https://medium.com/@benbjohnson/structuring-applications-in-go-3b04be4ff091) for a better explanation than I could ever do ha

**Comment by gschier** [comment]
Oh I see the problem. I had forgotten to push the latest version of the sample project. Should be up to date now. I think the out-of-sync problem can be easily avoided by not pushing everything into master all the time (once things stabilize).

Good point on the CLI. I think that, even though the CLI doesn't share any code with the framework, it's probably easier to manage both in one project for now.

**Comment by SpencerCDixon** [comment]
When using a github starter there is also the benefit that you could support different boilerplates.  I could imagine something like: `hemlock new myapp --starter=github.com/spencercdixon/h-boiler` that pulls that project instead of your default one.  Piggy back on `go get`'s API a bit. Pros and cons for sure!
